source/common/grpc/google_async_client_impl.cc

```diff
@@ -10,18 +10,56 @@
 namespace Envoy {
 namespace Grpc {
 
+GoogleAsyncClientThreadLocal::GoogleAsyncClientThreadLocal(Event::Dispatcher& dispatcher)
+    : dispatcher_(dispatcher),
+      completion_thread_(new Thread::Thread([this] { completionThread(); })) {}
+
+GoogleAsyncClientThreadLocal::~GoogleAsyncClientThreadLocal() {
+  // Force streams to shutdown and invoke TryCancel() to start the drain of
+  // pending op. If we don't do this, Shutdown() below can jam on pending ops.
+  // This is also required to satisfy the contract that once Shutdown is called,
+  // streams no longer queue any additional tags.
+  for (auto it = streams_.begin(); it != streams_.end();) {
+    (*it++)->resetStream();
+  }
+  cq_.Shutdown();
+  ENVOY_LOG(debug, "Joining completionThread");
+  completion_thread_->join();
+  ENVOY_LOG(debug, "Joined completionThread");
+  // Ensure that we have cleaned up all orphan streams, now that CQ is gone.
+  // Note, we expect that we're not in a dispatcher loop here, and participating
+  // in global teardown.
+  if (!streams_.empty()) {
+    dispatcher_.run(Event::Dispatcher::RunType::NonBlock);
```

Comment:
@mattklein123 I couldn't find a better way to do this off-the-to-of-my-head. Basically, we have this situation:

1. Per-silo TLS for GoogleAsyncClient is being torn down during server exit, as the slot gets blown away. We're on a given silo and issue `resetStream()` to the various streams still in existence.
2. Streams may have in-flight ops on their CQ thread that is buddy to the silo thread.
3. We shutdown the CQ thread and join it. There may still be in-flight completions that the CQ thread has posted back to the silo thread.
4. To avoid leaking memory (which gets picked up in tests) and bound lifetime of the stream cleanup, we need to execute some dispatcher events on the loop to cleanup up these inflight completions on the dispatcher.

The solution here is a bit weird, we let the dispatcher run a bit. However, this is likely happening after the usual dispatcher main loop finish for a worker, so seems like there might be dragons, and it's kind of special snowflake behavior.

We can maybe chat today when you visit Google a bit about this.

Reply:
Harvey and I just talked about this IRL. I think the plan is to not use dispatcher post for this but to implement posting functionality directly in this code which will make cleanup easier.